### ISSUE TITLE
plant translator

### ISSUE THREAD

**Issue by Plantman369** [Plantman369/test_01#1]
import kivy
from kivy.app import App
from kivy.uix.boxlayout import BoxLayout
from kivy.uix.button import Button
from kivy.uix.checkbox import CheckBox
from kivy.uix.label import Label
from kivy.uix.textinput import TextInput
from kivy.uix.filechooser import FileChooserIconView
from kivy.uix.scrollview import ScrollView
from kivy.uix.popup import Popup
from kivy.uix.spinner import Spinner
from kivy.uix.switch import Switch
import mido
import requests
import language_tool_python
import threading
from googletrans import Translator
import numpy as np
from sklearn.cluster import KMeans
from sklearn.metrics.pairwise import cosine_similarity
import collections
import nltk
from nltk.probability import ConditionalFreqDist
from nltk.util import bigrams, ngrams
import tensorflow as tf
from datetime import timedelta
import matplotlib.pyplot as plt
from kivy.core.audio import SoundLoader

# Get words from dictionary API
def get_words_from_api(query):
    url = f"https://api.dictionaryapi.dev/api/v2/entries/en/{query}"
    response = requests.get(url)
    if response.status_code == 200:
        data = response.json()
        words = [meaning['definitions'][0]['definition'] for meaning in data[0]['meanings']]
        return words
    return ["Unknown word"]

def detecter_accord(notes_en_cours):
    accords = {
        frozenset([60, 64, 67]): "C major", frozenset([62, 65, 69]): "D major",
        frozenset([64, 67, 71]): "E major", frozenset([65, 69, 72]): "F major",
        frozenset([67, 71, 74]): "G major", frozenset([69, 72, 76]): "A major",
        frozenset([71, 74, 77]): "B major", frozenset([60, 63, 67]): "C minor",
        frozenset([62, 65, 68]): "D minor", frozenset([64, 67, 70]): "E minor",
        frozenset([65, 68, 72]): "F minor", frozenset([67, 70, 74]): "G minor",
        frozenset([69, 72, 75]): "A minor", frozenset([71, 74, 77]): "B minor",
        frozenset([60, 63, 66]): "C diminished", frozenset([62, 65, 68]): "D diminished",
        frozenset([64, 67, 70]): "E diminished", frozenset([65, 68, 71]): "F diminished",
        frozenset([67, 70, 73]): "G diminished", frozenset([69, 72, 75]): "A diminished",
        frozenset([71, 74, 77]): "B diminished", frozenset([60, 64, 68]): "C augmented",
        frozenset([62, 66, 70]): "D augmented", frozenset([64, 68, 72]): "E augmented",
        frozenset([65, 69, 73]): "F augmented", frozenset([67, 71, 75]): "G augmented",
        frozenset([69, 73, 77]): "A augmented", frozenset([71, 75, 79]): "B augmented",
    }
    return accords.get(frozenset(notes_en_cours), "Unknown chord")

def convertir_temps(temps_en_ms):
    return str(timedelta(milliseconds=temps_en_ms))

def corriger_texte(texte, langue='fr'):
    tool = language_tool_python.LanguageTool(langue)
    matches = tool.check(texte)
    texte_corrige = language_tool_python.utils.correct(texte, matches)
    return texte_corrige

def analyser_motif_markov(sequences):
    return ConditionalFreqDist(bigrams(sequences))

def analyser_motif_ngram(sequences, n=3):
    return collections.Counter(ngrams(sequences, n))

def analyser_motif_similarity_cosine(sequences):
    return cosine_similarity(np.array(sequences).reshape(-1, 1))

def analyser_motif_clustering(sequences, n_clusters=3):
    return KMeans(n_clusters=n_clusters, random_state=0).fit(np.array(sequences).reshape(-1, 1))

def convertir_midi_en_mots_realtime(sequences, use_markov=False, use_ngram=False, use_cosine=False, use_clustering=False):
    texte = []
    if use_markov:
        markov_model = analyser_motif_markov(sequences)
        texte.append(f"Markov Chain Model: {markov_model}")
    if use_ngram:
        ngram_model = analyser_motif_ngram(sequences)
        texte.append(f"N-gram Model: {ngram_model}")
    if use_cosine:
        cosine_similarity_matrix = analyser_motif_similarity_cosine(sequences)
        texte.append(f"Cosine Similarity: {cosine_similarity_matrix}")
    if use_clustering:
        clustering_model = analyser_motif_clustering(sequences)
        texte.append(f"Clustering Model: {clustering_model}")
    return texte

def save_results_to_file(results):
    content = BoxLayout(orientation='vertical')
    file_chooser = FileChooserIconView()
    save_button = Button(text="Save", size_hint_y=None, height=50)
    content.add_widget(file_chooser)
    content.add_widget(save_button)

    popup = Popup(title='Save Results', content=content, size_hint=(0.9, 0.9))

    def on_save(instance):
        if file_chooser.selection:
            file_path = file_chooser.selection[0]
            with open(file_path, 'w') as file:
                file.write(results)
            popup.dismiss()

    save_button.bind(on_release=on_save)
    popup.open()

def plot_graphs(sequences):
    plt.figure(figsize=(10, 6))
    plt.subplot(2, 1, 1)
    plt.hist(sequences, bins=range(min(sequences), max(sequences) + 1))
    plt.title('Histogram of Notes')
    plt.xlabel('Note')
    plt.ylabel('Frequency')

    plt.subplot(2, 1, 2)
    plt.plot(sequences, linestyle='-', marker='o')
    plt.title('Sequence of Notes')
    plt.xlabel('Index')
    plt.ylabel('Note')

    plt.tight_layout()
    plt.show()

class MidiApp(App):
    def build(self):
        self.title = "Plant Music Interpreter"
        self.root = BoxLayout(orientation='vertical')
        self.markov_var = False
        self.ngram_var = False
        self.cosine_var = False
        self.clustering_var = False

        # File chooser
        self.file_chooser = FileChooserIconView(multiselect=True)
        self.root.add_widget(self.file_chooser)

        # Checkboxes for analytical features
        self.checkbox_layout = BoxLayout(orientation='horizontal', size_hint_y=None, height=50)
        self.checkbox_layout.add_widget(CheckBox(size_hint=(None, None), size=(50, 50), on_active=self.on_markov_toggle))
        self.checkbox_layout.add_widget(Label(text="Markov Chain"))
        self.checkbox_layout.add_widget(CheckBox(size_hint=(None, None), size=(50, 50), on_active=self.on_ngram_toggle))
        self.checkbox_layout.add_widget(Label(text="N-grams"))
        self.checkbox_layout.add_widget(CheckBox(size_hint=(None, None), size=(50, 50), on_active=self.on_cosine_toggle))
        self.checkbox_layout.add_widget(Label(text="Cosine Similarity"))
        self.checkbox_layout.add_widget(CheckBox(size_hint=(None, None), size=(50, 50), on_active=self.on_clustering_toggle))
        self.checkbox_layout.add_widget(Label(text="K-Means Clustering"))
        self.root.add_widget(self.checkbox_layout)

        # Result text area
        self.result_text = TextInput(size_hint_y=None, height=500, readonly=True)
        scroll_view = ScrollView(size_hint=(1, None), size=(400, 400))
        scroll_view.add_widget(self.result_text)
        self.root.add_widget(scroll_view)

        # Buttons
        self.open_file_button = Button(text="Open MIDI File(s)", size_hint_y=None, height=50, on_release=self.open_file)
        self.root.add_widget(self.open_file_button)

        self.save_results_button = Button(text="Save Results", size_hint_y=None, height=50, on_release=self.save_results)
        self.root.add_widget(self.save_results_button)

        self.plot_graphs_button = Button(text="Plot Graphs", size_hint_y=None, height=50, on_release=self.plot_graphs)
        self.root.add_widget(self.plot_graphs_button)

        # Language selection
        self.language_spinner = Spinner(
            text='fr',
            values=('en', 'fr', 'es', 'de', 'it', 'pt'),
            size_hint=(None, None),
            size=(100, 50)
        )
        self.root.add_widget(Label(text="Language"))
        self.root.add_widget(self.language_spinner)

        # Dark mode switch
        self.dark_mode_switch = Switch(active=False)
        self.dark_mode_switch.bind(active=self.on_dark_mode_toggle)
        self.root.add_widget(Label(text="Dark Mode"))
        self.root.add_widget(self.dark_mode_switch)

        self.notification_sound = SoundLoader.load('notification.wav')
        return self.root

    def on_markov_toggle(self, instance, value):
        self.markov_var = value

    def on_ngram_toggle(self, instance, value):
        self.ngram_var = value

    def on_cosine_toggle(self, instance, value):
        self.cosine_var = value

    def on_clustering_toggle(self, instance, value):
        self.clustering_var = value

    def open_file(self, instance):
        file_paths = self.file_chooser.selection
        if file_paths:
            for file_path in file_paths:
                threading.Thread(target=self.process_midi_file, args=(file_path,)).start()

    def process_midi_file(self, file_path):
        translator = Translator()
        try:
            mid = mido.MidiFile(file_path)
            sequences = []
            notes_en_cours = []
            texte = []

            for track in mid.tracks:
                for message in track:
                    if message.type in ['note_on', 'note_off']:
                        sequences.append(message.note)
                        if message.type == 'note_on' and message.velocity > 0:
                            notes_en_cours.append(message.note)
                        elif message.type == 'note_off':
                            if message.note in notes_en_cours:
                                notes_en_cours.remove(message.note)

                        accord = detecter_accord(notes_en_cours)
                        description = f"Type: {message.type}, Note: {message.note}, Chord: {accord}"
                        description_translated = translator.translate(description, dest=self.language_spinner.text).text
                        texte.append(description_translated)

            self.result_text.text += "\n".join(texte) + "\n"
            self.sequences = sequences
        except Exception as e:
            self.result_text.text += f"Error processing {file_path}: {str(e)}\n"

    def save_results(self, instance):
        save_results_to_file(self.result_text.text)

    def plot_graphs(self, instance):
        if hasattr(self, 'sequences'):
            plot_graphs(self.sequences)

    def on_dark_mode_toggle(self, instance, value):
        if value:
            self.root.background_color = (0, 0, 0, 1)
            for child in self.root.children:
                if isinstance(child, Label):
                    child.color = (1, 1, 1, 1)
        else:
            self.root.background_color = (1, 1, 1, 1)
            for child in self.root.children:
                if isinstance(child, Label):
                    child.color = (0, 0, 0, 1)

if __name__ == '__main__':
    MidiApp().run()
